test/compiler/inference.jl

```diff
@@ -4937,10 +4937,19 @@ g() = empty_nt_values(Base.inferencebarrier(Tuple{}))
 # This is somewhat sensitive to the exact recursion level that inference is willing to do, but the intention
 # is to test the case where inference limited a recursion, but then a forced constprop nevertheless managed
 # to terminate the call.
+@newinterp RecurseInterpreter
+function CC.bail_out_const_call(interp::RecurseInterpreter, result::CC.MethodCallResult,
+                                si::CC.StmtInfo, sv::CC.AbsIntState)
+    if result.rt isa CC.LimitedAccuracy
+        return false # allow constprop to recurse into unresolved cycles
+    end
+    return @invoke CC.bail_out_const_call(interp::CC.AbstractInterpreter, result::CC.MethodCallResult,
+                                          si::CC.StmtInfo, sv::CC.AbsIntState)
+end
 Base.@constprop :aggressive type_level_recurse1(x...) = x[1] == 2 ? 1 : (length(x) > 100 ? x : type_level_recurse2(x[1] + 1, x..., x...))
 Base.@constprop :aggressive type_level_recurse2(x...) = type_level_recurse1(x...)
 type_level_recurse_entry() = Val{type_level_recurse1(1)}()
-@test Base.return_types(type_level_recurse_entry, ()) |> only == Val{1}
+@test Base.infer_return_type(type_level_recurse_entry, (); interp=RecurseInterpreter()) == Val{1}
```

Comment:
@Keno This PR prevents const-prop' in the presence of unresolvable cycles, regardless of whether const-prop' is enforced. Consequently, it causes the test cases from #48045 to fail. For interpreters requiring optimization against heavy recursions, like Diffractor, do you think it would fine to make them add explicit overloads to permit const-prop' for cycles as like `RecurseInterpreter` here?

Reply:
I don't fully know. I'm now entirely happy with the recursion story with respect to Diffractor anyway, so I think that needs a larger solution. As long as this doesn't end up causing us immediate problems, I think it's fine.

Reply:
In that case, I'll move forward with this plan. I'll implement the necessary adjustments in Diffractor that would be similar to what `RecurseInterpreter` does here.